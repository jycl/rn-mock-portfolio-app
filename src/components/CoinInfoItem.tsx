import React from 'react';
import { View, Text, Image, StyleSheet, Pressable } from 'react-native';
import RowPair from './RowPair';
import { convertToCurrency } from '../util';
import Assets from '../assets';

type PropsType = {
  name: string;
  price: string;
  quantity: string;
  change24hour: number;
  onPress?: () => void;
};

/**
 * renderItem component for FlatList in Portfolio screen for
 * cryptocurrencies.
 */
export default function CoinInfoItem({
  name,
  price,
  quantity,
  change24hour,
  onPress,
}: PropsType) {
  const change = `${Number(change24hour) > 0 ? '+' : '-'}${change24hour}%`;
  const key = name.toLowerCase();
  return (
    <Pressable
      style={styles.container}
      onPress={onPress}
      testID={`testID_${name}`}>
      <View style={styles.coinContainer}>
        {!!Assets[key] && <Image style={styles.img} source={Assets[key]} />}
        <Text style={styles.text}>{name}</Text>
      </View>
      <RowPair
        containerStyle={styles.priceContainer}
        upper={convertToCurrency(price)}
        lower={change}
        lowerColor="green"
      />
      <RowPair
        containerStyle={styles.valueContainer}
        upper={convertToCurrency(Number(price) * Number(quantity))}
        lower={quantity}
        lowerColor="grey"
      />
    </Pressable>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    minHeight: 64,
    alignItems: 'center',
  },
  text: {
    color: 'white',
    fontSize: 18,
  },
  coinContainer: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  valueContainer: {
    flex: 1,
    justifyContent: 'flex-start',
    paddingRight: 16,
  },
  priceContainer: {
    flex: 1,
    justifyContent: 'flex-start',
  },
  img: {
    width: 24,
    height: 24,
    margin: 12,
  },
});
