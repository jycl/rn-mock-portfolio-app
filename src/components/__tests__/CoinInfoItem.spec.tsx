import React from 'react';
import { create } from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';
import CoinInfoItem from '../CoinInfoItem';

describe('CoinInfoItem component', () => {
  const mock = {
    name: 'BTC',
    price: '12.1201',
    quantity: '201',
    change24hour: 4.99,
    onPress: jest.fn(),
  };
  it('should match snapshot render with defined props for positive change', () => {
    const coinRender = create(<CoinInfoItem {...mock} />);
    expect(coinRender.toJSON()).toMatchSnapshot();
  });

  it('should match snapshot render with defined props for negative change', () => {
    const localMock = {
      ...mock,
      change24hour: -3.25,
    };
    const coinRender = create(<CoinInfoItem {...localMock} />);
    expect(coinRender.toJSON()).toMatchSnapshot();
  });

  it('should invoke onPress callback on press component', () => {
    const { getByTestId } = render(<CoinInfoItem {...mock} />);
    fireEvent.press(getByTestId('testID_BTC'));
    expect(mock.onPress).toBeCalled();
  });
});
