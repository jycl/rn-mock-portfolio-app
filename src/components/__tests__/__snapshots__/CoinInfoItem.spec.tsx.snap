// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinInfoItem component should match snapshot render with defined props for negative change 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "minHeight": 64,
    }
  }
  testID="testID_BTC"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/btc_logo.png",
        }
      }
      style={
        Object {
          "height": 24,
          "margin": 12,
          "width": 24,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
        }
      }
    >
      BTC
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-end",
          "flexDirection": "column",
        },
        Object {
          "flex": 1,
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
        }
      }
    >
      $12.12
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          Object {
            "color": "green",
          },
        ]
      }
    >
      --3.25%
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-end",
          "flexDirection": "column",
        },
        Object {
          "flex": 1,
          "justifyContent": "flex-start",
          "paddingRight": 16,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
        }
      }
    >
      $2436.14
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          Object {
            "color": "grey",
          },
        ]
      }
    >
      201
    </Text>
  </View>
</View>
`;

exports[`CoinInfoItem component should match snapshot render with defined props for positive change 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "minHeight": 64,
    }
  }
  testID="testID_BTC"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/btc_logo.png",
        }
      }
      style={
        Object {
          "height": 24,
          "margin": 12,
          "width": 24,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
        }
      }
    >
      BTC
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-end",
          "flexDirection": "column",
        },
        Object {
          "flex": 1,
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
        }
      }
    >
      $12.12
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          Object {
            "color": "green",
          },
        ]
      }
    >
      +4.99%
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-end",
          "flexDirection": "column",
        },
        Object {
          "flex": 1,
          "justifyContent": "flex-start",
          "paddingRight": 16,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
        }
      }
    >
      $2436.14
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          Object {
            "color": "grey",
          },
        ]
      }
    >
      201
    </Text>
  </View>
</View>
`;
